src/routes/Project/actions/TrackActionHandler.js

```diff
@@ -47,101 +82,168 @@ const ACTION_HANDLERS = {
         ...state,
         [oldClip.track]: oldTrack,
         [newClip.track]: {
           ...state[newClip.track],
-          [newClip.startTime]: id,
-          [newClip.endTime]: id,
           clips: {
             ...state[newClip.track].clips,
             [id]: newClip
           }
         }
       };
 
-      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track]);
+      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track].clips);
     } else {
+      if (collision) {
+        newClip.track = oldClip.track;
+      }
+
+      if (!oldClip.ghostClip) {
+        adjustForCollisions(newClip, state[oldClip.track], isMovingRight);
+      } else if (oldClip.ghostClip && !isCollision(newClip, state[oldClip.track])) {
+        newClip.ghostClip = false;
+      }
+
       newTracks = {
         ...state,
-        [newClip.track]: {
-          ...state[newClip.track],
-          [newClip.startTime]: id,
-          [newClip.endTime]: id,
+        [oldClip.track]: {
+          ...state[oldClip.track],
           clips: {
-            ...state[newClip.track].clips,
+            ...state[oldClip.track].clips,
             [id]: newClip
           }
         }
       };
 
-      delete newTracks[newClip.track][oldClip.startTime];
-      delete newTracks[newClip.track][oldClip.endTime];
-      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track]);
+      newTracks[oldClip.track].edgeClip = findEdgeClip(newTracks[oldClip.track].clips);
     }
 
     return newTracks;
+  },
+
+  [ProjectConstants.DELETE_CLIP]: (state, action) => {
+    const clip = action.payload;
+    let newClips = { ...state[clip.track].clips };
+    delete newClips[clip.id];
+
+    return {
+      ...state,
+      [clip.track]: {
+        ...state[clip.track],
+        clips: newClips,
+        edgeClip: findEdgeClip(newClips)
+      }
+    };
   }
 };
 
 export default ACTION_HANDLERS;
 
 /* HELPERS */
-function findEdgeClip(track) {
+function findEdgeClip(clips) {
   let max;
-  for (let time in track) {
+  let maxClip;
+  for (let clipId in clips) {
+    let clip = clips[clipId];
+    let time = clip.endTime;
     time = Number(time);
     if (typeof time === 'number' && !Number.isNaN(time)) {
-      if (!max || time > max) max = time;
+      if (!max || time > max) {
+        max = time;
+        maxClip = clip;
+      }
     }
   }
 
   if (!max) return null;
-  return track.clips[track[max]];
+  return maxClip;
+}
+
+// handle collision avoidance. not very efficient but should work.
+function adjustForCollisions(newClip, track, isMovingRight) {
+  const times = _(track.clips).keys()
+    .reduce((acc, currTrackId) => {
+      let currTrack = track.clips[currTrackId];
+      acc.push({ time: Number(currTrack.startTime), id: currTrackId });
+      acc.push({ time: Number(currTrack.endTime), id: currTrackId });
+      return acc;
+    }, [])
+    .sort(sortAscendingTimeEntries);
+
+  for (let { id } of times) {
```

Comment:
this and the above should be refactored to just iterate through all clips instead of the time indirection:
`for (timeClip of_.omitBy(track.clips, {id: newClip.id}))`